narwhals/dtypes.py

```diff
@@ -460,35 +461,62 @@ class Categorical(DType):
     """
 
 
+class _DelayedCategories:
+    """Store callable which produces tupleified version of Enum's categories.
+
+    !!! warning
+        This class is not meant to be instantiated directly and exists
+        for internal usage only.
+    """
+
+    def __init__(self, get_categories: Callable[[], tuple[str, ...]]) -> None:
+        self.get_categories = get_categories
+
```

Comment:
This way we can avoid adding `categories: _DelayedCategories` 

```suggestion

    def __iter__(self) -> Iterator[str]:
        yield from self.get_categories()

```